sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,518 @@
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return VoidCoder.of();
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+  }
+
+  /**
+   * POJO used to store MQTT message and its timestamp.
+   */
+  private static class MqttMessageWithTimestamp {
+
+    private MqttMessage message;
+    private Instant timestamp;
+
+    public MqttMessageWithTimestamp(MqttMessage message, Instant timestamp) {
+      this.message = message;
+      this.timestamp = timestamp;
+    }
+
+    public MqttMessage getMessage() {
+      return message;
+    }
+
+    public void setMessage(MqttMessage message) {
+      this.message = message;
+    }
+
+    public Instant getTimestamp() {
+      return timestamp;
+    }
+
+    public void setTimestamp(Instant timestamp) {
+      this.timestamp = timestamp;
+    }
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader<byte[]> {
+
+    private final UnboundedMqttSource source;
+
+    private MqttClient client;
+    private byte[] current;
+    private Instant currentTimestamp;
+    private BlockingQueue<MqttMessageWithTimestamp> queue;
+
+    private UnboundedMqttReader(UnboundedMqttSource source) {
+      this.source = source;
+      this.current = null;
+      this.queue = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOGGER.debug("Starting MQTT reader");
+      Read spec = source.spec;
+      try {
+        client = spec.mqttConnectionConfiguration().getClient();
+        client.subscribe(spec.mqttConnectionConfiguration().topic());
+        client.setCallback(new MqttCallback() {
+          @Override
+          public void connectionLost(Throwable cause) {
+            LOGGER.warn("MQTT connection lost", cause);
+            try {
+              close();
+            } catch (Exception e) {
+              // nothing to do
+            }
+          }
+
+          @Override
+          public void messageArrived(String topic, MqttMessage message) throws Exception {
+            LOGGER.trace("Message arrived");
+            queue.put(new MqttMessageWithTimestamp(message, Instant.now()));
+          }
+
+          @Override
+          public void deliveryComplete(IMqttDeliveryToken token) {
+            // nothing to do
+          }
+        });
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      LOGGER.debug("Taking from the pending queue ({})", queue.size());
+      try {
+        MqttMessageWithTimestamp message = queue.take();
```

Comment:
I would recommend to set some timeout on take(), so that if new messages completely stop arriving on the queue, the runner will still be able to finish the current bundle.